Create SandboxTest script

felipecrs opened the pull request:
**PS:** check [this comment](https://github.com/microsoft/winget-pkgs/pull/827#issuecomment-656422758) if you want to use it before it gets merged.

# Description

A new PowerShell script to test the installation of any Manifest in Windows Sandbox.

# Usage

```sh-session
$ Get-Help .\Tools\SandboxTest.ps1
SandboxTest.ps1 [[-Manifest] <string>] [[-Script] <scriptblock>] [-MapFolder <string>] [<CommonParameters>]
```

## Examples

```sh-session
$ .\Tools\SandboxTest.ps1 .\Tools\SandboxTest.ps1 .\manifests\GitHub\cli\0.10.1.yaml { gh --version }
--> Validating Manifest
Manifest validation succeeded.

--> Downloading dependencies

--> Starting Windows Sandbox, and:
    - Mounting the following directories:
      - C:\Users\felip\AppData\Local\Temp\SandboxTest as read-only
      - C:\Users\felip\Repos\winget-pkgs as read-and-write
    - Installing WinGet
    - Installing the Manifest .\manifests\GitHub\cli\0.10.1.yaml
    - Refreshing environment variables
    - Running the following script:

{
 gh --version
}
```

A GIF is worth a thousand words.

![U6ho8iV7OL](https://user-images.githubusercontent.com/29582865/90988322-56922800-e568-11ea-81f9-a32f191c35f9.gif)

# To-do list

- [x] Download `winget` `.Appx` files on the fly, including the dependencies such as the `VCLibs` **(and remove them from git, of course)**

- [x] Validate Manifest file with `winget validate` before pushing to Sandbox test

- [x] Install `winget` in Sandbox 

- [x] Run `winget install -m` with the provided manifest file in Sandbox

- [x] Add ability to run the custom script as the last step in Sandbox

- [x] Automatically refresh the environment variables after installing the Manifest

- [x] Close Windows Sandbox automatically if needed

- [x] Provide a way to specify which folder to mount, default is the current

Fixes https://github.com/microsoft/winget-cli/issues/253, despite personally I would love to see this integrated into `winget-cli` itself, so at least users would not need to clone the repository before testing.

###### Microsoft Reviewers: [Open in CodeFlow](http://wpcp.azurewebsites.net/CodeFlowProtocolProxy2.php?pullrequest=https://github.com/microsoft/winget-pkgs/pull/827)

felipecrs committed:
Create SandboxTest script

felipecrs committed:
Add missing appx dependencies

felipecrs committed:
Clear temp folder

felipecassiors commented:
I need help with the `winget validate`, it always returns 0 as status code so I'm not sure how can I catch errors in it. Does anyone have any suggestions?

felipecassiors commented:
I also need to know where I can get the VCLibs and VCLibsUWP `.Appx` files, which need to be installed as dependencies for the `DesktopAppInstaller`. Does anyone know? 

I tried the `.Appx` provided by [this package](https://www.microsoft.com/en-us/download/details.aspx?id=53175), but they are not up-to-date and does not meet the requirements for the `DesktopAppInstaller`.

felipecrs committed:
Merge branch 'master' of https://github.com/microsoft/winget-pkgs into sandbox-test

chausner commented:
Very cool. Thanks for sharing this prototype script!

> I need help with the `winget validate`, it always returns 0 as status code so I'm not sure how can I catch errors in it. Does anyone have any suggestions?

Good point. `winget validate` should indeed exit with a non-zero error code if validation fails. For now, you could just parse the output and look for the string "Manifest validation succeeded.".

chausner commented:
> I also need to know where I can get the VCLibs and VCLibsUWP `.Appx` files, which need to be installed as dependencies for the `DesktopAppInstaller`. Does anyone know?
> 
> I tried the `.Appx` provided by [this package](https://www.microsoft.com/en-us/download/details.aspx?id=53175), but they are not up-to-date and does not meet the requirements for the `DesktopAppInstaller`.

People wanting to install Windows Terminal outside of the Store have run into the same issue. An up-to-date version of the dependencies is available in the Store: https://www.microsoft.com/store/productId/9NBLGGH4RV3K

Unfortunately, there does not seem to be a way to automate its installation. A manual workaround is described [here](https://github.com/microsoft/terminal/issues/3097#issuecomment-615373769).

felipecrs committed:
Validate manifest before pushing to sandbox

felipecassiors commented:
> Good point. `winget validate` should indeed exit with a non-zero error code if validation fails (I created issue [microsoft/winget-cli#312](https://github.com/microsoft/winget-cli/issues/312)). For now, you could just parse the output and look for the string "Manifest validation succeeded.".

Good idea! I did it.

felipecassiors commented:
> People wanting to install Windows Terminal outside of the Store have run into the same issue. An up-to-date version of the dependencies is available in the Store: https://www.microsoft.com/store/productId/9NBLGGH4RV3K

I used https://store.rg-adguard.net/ to download the `.Appx` from the Store. However, despite the links for the `.Appx` there are directly from Microsoft, they expire, so we can't use them for downloading. 

We can still upload these two `.Appx` somewhere, such as in GitHub releases, but I don't if this is allowed or not.

> Unfortunately, there does not seem to be a way to automate its installation. A manual workaround is described [here](https://github.com/microsoft/terminal/issues/3097#issuecomment-615373769).

I had read this before, but I didn't get what was the manual workaround. It seems that the Windows Terminal installer now ships the dependencies builtin, and if that's true, this could solve the problem. But one open point is that someone said the installer just tells the Store to download the dependencies automatically, if that's the case this wouldn't help since the Windows Sandbox doesn't have access to the Store.

chausner committed:
Fix invalid character in Sandbox.ps1

The dash character in -ExecutionPolicy was a special character which broke the command.

felipecrs committed:
Merge branch 'master' of https://github.com/microsoft/winget-pkgs into sandbox-test

chausner commented:
> I also need to know where I can get the VCLibs and VCLibsUWP `.Appx` files, which need to be installed as dependencies for the `DesktopAppInstaller`. Does anyone know?

Since the script requires winget to be installed, those VCLibs dependencies should also be guaranteed to be present on the host system. I tried building the appx files from C:\ProgramFiles\WindowsApps\Microsoft.VCLibs.. by simply zipping the folders but for some reason the resulting appx files become corrupt. Even taking a valid appx, unzipping and zipping it again, makes them corrupt. Still trying to figure out why.

felipecassiors commented:
@chausner Right. But did you tried with the files in this PR? The binaries are in the branch until I solve that.

chausner commented:
> @chausner Right. But did you tried with the files in this PR? The binaries are in the branch until I solve that.

Yes, the current solution works fine (apart from this minor issue https://github.com/felipecassiors/winget-pkgs/pull/1). I was just checking if there is an alternative way to obtain the dependencies.

felipecrs committed:
Download dependencies on the fly

felipecrs committed:
remove vscode launch file

felipecrs committed:
Add newline in gitignore

felipecassiors commented:
@chausner Can you please take a look now? I removed the binaries from the branch, and now the script is able to download them if needed.

The only thing missing now is the URLs for the VCLibs dependencies (not a problem if we can upload them to somewhere).

felipecrs committed:
Fix when there is no temp folder

felipecassiors commented:
This PR is ready for review.

![YutrU9BLac](https://user-images.githubusercontent.com/29582865/83085672-0556c300-a063-11ea-8d9b-055491877fde.gif)

felipecrs committed:
Add Test in Sandbox to README

felipecrs committed:
Add some comments and enhance some messages

chausner commented on Tools/SandboxTest.ps1:
```diff
@@ -0,0 +1,110 @@
+    </MappedFolder>
+  </MappedFolders>
+  <LogonCommand>
+  <Command>PowerShell Start-Process PowerShell -WorkingDirectory '$tempFolderInSandbox' -ArgumentList '-ExecutionPolicy Bypass -NoExit -File $bootstrapPs1FileName'</Command>
```
I think it should be possible to get rid of the nested invocation of PowerShell here. Instead of specifying the working directory on the command-line, you could put `Set-Location $PSScriptRoot` at the beginning of the bootstrap script. Then the command-line could be simply
```
PowerShell -ExecutionPolicy Bypass -NoExit -File $bootstrapPs1FileName
````

felipecassiors replied:
`Set-Location $PSScriptRoot` would increase the complexity:
1. A new line in `Bootstrap.ps1`
2. Have to pass the full path in Sandbox for `Bootstrap.ps1` in the `-File` PowerShell parameter

felipecassiors replied:
I tried many combinations, and the current one is the only one that opens PowerShell in the foreground and keeps its window open.

In your example, the script runs in the background.

chausner commented on Tools/SandboxTest.ps1:
```diff
@@ -0,0 +1,110 @@
+    # This downloads the file
+    Write-Host "Downloading $($dependency.url) ..."
+    try { 
+      $WebClient.DownloadFile($dependency.url, $dependency.file) 
```
Could be written a bit shorter without instantiating WebClient by using `Invoke-WebRequest -Uri $dependency.url -OutFile $dependency.file`.

felipecassiors replied:
I chose `WebClient` because:
1. `YamlCreate.ps1` also uses.
2. This [site](https://blog.jourdant.me/post/3-ways-to-download-files-with-powershell) says the performance is way better.
Are you sure you want me to change?

chausner replied:
That's good to know, I was not aware of the performance issues of Invoke-WebRequest (which, ironically, appear to be solely due to the progress bar). While you can disable the progress bar, I think WebClient is also fine, also since it's used by `YamlCreate.ps1`.

chausner commented on Tools/SandboxTest.ps1:
```diff
@@ -0,0 +1,110 @@
+# Parse Arguments
+
+param ([string] $Manifest = $(throw "The Manifest parameter is required."))
```
I think `[Parameter(Mandatory=$true)]` is the canonical way to mark parameters as required.

felipecassiors replied:
Awesome!!! Great to know!

felipecassiors replied:
Implemented!

felipecassiors replied:
But sadly this made the script to always print a blank newline at the beginning. I didn't find how to prevent that.

chausner commented:
The script should check at the beginning whether Windows Sandbox is enabled as a Windows feature and if not, print out a helpful error message. While there is `Get-WindowsOptionalFeature -Online -FeatureName "Containers-DisposableClientVM"` to check the installation state, it requires elevated rights. So it's probably better to simply check with `Test-Path "$env:windir\System32\WindowsSandbox.exe"`.

felipecrs committed:
Properly accept mandatory parameter

felipecrs committed:
Check if Windows Sandbox exist or not

felipecassiors commented:
> The script should check at the beginning whether Windows Sandbox is enabled as a Windows feature and if not, print out a helpful error message. While there is `Get-WindowsOptionalFeature -Online -FeatureName "Containers-DisposableClientVM"` to check the installation state, it requires elevated rights. So it's probably better to simply check with `Test-Path "$env:windir\System32\WindowsSandbox.exe"`.

Implemented.

chausner commented on Tools/SandboxTest.ps1:
```diff
@@ -16,6 +16,18 @@ if ($validationResult -like '*Manifest validation failed.*') {
+https://docs.microsoft.com/en-us/windows/security/threat-protection/windows-sandbox/windows-sandbox-overview
+  
+You can run the following command in an elevated PowerShell for enabling Windows Sandbox:
+Get-WindowsOptionalFeature -Online -FeatureName 'Containers-DisposableClientVM'
```
The command to enable it would be `Enable-WindowsOptionalFeature`. And there's a small typo: avaiable.

felipecassiors replied:
That's right. Sorry.

felipecassiors replied:
That's right. Sorry.

felipecrs committed:
Fix Sandbox install snippet and typo

denelon commented:
@felipecassiors I'll reach out to a couple of people to review this.

zachcarp commented on Tools/SandboxTest.ps1:
```diff
@@ -0,0 +1,125 @@
+'@ -ErrorAction Stop
+}
+
+# Set dependencies
```
Just a thought - with these dependencies hard-coded, we will need to update the script each time we release. Have you thought about downloading the files from the latest api? (https://api.github.com/repos/microsoft/winget-cli/releases/latest). Or you could use it as a prompt to notify a user of the script that a new client is available.

felipecassiors replied:
Yes, I thought about that. I was wondering to wait for the second release of `winget` to catch the pattern (I don't know if the file name is going to change or not). But that's good, I'll implement it now.

zachcarp replied:
I believe @KevinLaMS and @denelon will primarily be taking point on deciding when to release an update (please correct me if I am wrong). Unless there is a good reason for having the _8wekyb3d8bbwe, we could probably remove that from the future releases and just go with Microsoft.DestkopAppInstaller.appxbundle

felipecassiors replied:
If `8wekyb3d8bbwe` is immutable, this wouldn't be a problem.

chausner replied:
That is the Microsoft PublisherId. It's the same for all apps published by Microsoft and should not change.

felipecassiors replied:
GitHub Releases API doesn't provide a hash for the assets, and without a hash, we can't verify if the existing binary is valid or not for triggering the download again. I think we should publish the hashes in some other place, maybe in the description of the release for this to work.

felipecassiors replied:
@zachcarp any opinion?

zachcarp replied:
The hash does present a problem. I see 3 alternatives.

1. Leave it as is for now, and revisit this after we figure out the other major issues. This lets us get this merged sooner.
2. Publish the hash somewhere. Your idea of putting it in the description works, but we will need to wait for another release for that to happen.
3. While not ideal, it at least deserves a mention. You can download the package without a hash, and hash it in your script. This allows you to move forward now, and guarantee fidelity when loading to the sandbox, but does nothing about ensuring fidelity when consuming from GitHub.

With those in mind, I am inclined to go with the first one for now. When we get another release, we can look at revisiting this. That also lets us focus on the dependency issue now.

felipecassiors replied:
I also agree with option one for now.

KevinLaMS commented on Tools/SandboxTest.ps1:
```diff
@@ -0,0 +1,125 @@
+
+$vcLibs = @{
+  fileName = 'Microsoft.VCLibs.140.00_14.0.27810.0_x64__8wekyb3d8bbwe.Appx'
+  url      = 'https://github.com/felipecassiors/winget-pkgs/raw/da8548d90369eb8f69a4738dc1474caaffb58e12/Tools/SandboxTest_Temp/Microsoft.VCLibs.140.00_14.0.27810.0_x64__8wekyb3d8bbwe.Appx'
```
I need to see if we can find a more official release path.  I will reach out to VS.

felipecassiors replied:
Sure! I used my GitHub as a workaround/placeholder.

KevinLaMS commented on Tools/SandboxTest.ps1:
```diff
@@ -0,0 +1,125 @@
+
+$vcLibsUwp = @{
+  fileName = 'Microsoft.VCLibs.140.00.UWPDesktop_14.0.27810.0_x64__8wekyb3d8bbwe.Appx'
+  url      = 'https://raw.githubusercontent.com/felipecassiors/winget-pkgs/da8548d90369eb8f69a4738dc1474caaffb58e12/Tools/SandboxTest_Temp/Microsoft.VCLibs.140.00.UWPDesktop_14.0.27810.0_x64__8wekyb3d8bbwe.Appx'
```
I need to see if we can find a more official release path.  I will reach out to VS.

felipecassiors replied:
I'll mark this conversation as closed because we have another open for the same matter.

zachcarp commented on Tools/SandboxTest.ps1:
```diff
@@ -0,0 +1,125 @@
+
+$vcLibs = @{
+  fileName = 'Microsoft.VCLibs.140.00_14.0.27810.0_x64__8wekyb3d8bbwe.Appx'
+  url      = 'https://github.com/felipecassiors/winget-pkgs/raw/da8548d90369eb8f69a4738dc1474caaffb58e12/Tools/SandboxTest_Temp/Microsoft.VCLibs.140.00_14.0.27810.0_x64__8wekyb3d8bbwe.Appx'
```
We should make sure dependencies are coming from an official source. If you deleted your fork or make alterations in the future, this could fail start to fail.

felipecassiors replied:
For sure! As I said I'm just using it as a placeholder.

felipecassiors replied:
I'll mark this conversation as closed, because we have another open for the same matter.

felipecassiors commented:
@KevinLaMS @zachcarp 

About the source for the URLs, let me elaborate. I can think of different solutions:

1. Include the dependencies in the `DesktopAppInstaller` package, as it seems they did in Windows Terminal, check [here](https://github.com/microsoft/terminal/issues/3097). But I'm not sure how that works, some people said it just tells Microsoft Store to install the dependencies, if that's the case, it won't solve since Windows Sandbox has no access to the Store.

2. Find the official source for the dependencies. Beware that https://www.microsoft.com/en-us/download/details.aspx?id=53175 can't be used, because of the version of the `.Appx` provided with the package is older than the required by `DesktopAppInstaller`, and also, I could not find the `UWPDesktop` variant of `VCLibs` on it.

3. Upload by ourselves to somewhere, it could be side by side with the `DesktopAppInstaller` package GitHub release for example.

Just to say, I got the `.Appx` for the dependencies using this third-party tool [here](https://store.rg-adguard.net/).

felipecassiors commented:
I just checked, and approach 1 would work. I tried to install Windows Terminal `.msixbundle` in the Windows Sandbox and there was no problems about the dependencies. I'll open a feature request in `winget-cli` for that.

MarNicGit commented:
@felipecassiors Does your script also allow me to check if the installed program will start? Right now I do everything manually in the win10 dev environment (that I reset to a checkpoint afterwards) and after installation I check if the application actually runs.

felipecassiors commented:
> @felipecassiors Does your script also allow me to check if the installed program will start? Right now I do everything manually in the win10 dev environment (that I reset to a checkpoint afterwards) and after installation I check if the application actually runs.

Yes, after the installation the Sandbox remains open. You can check the installation logs and open your program as well.

felipecrs committed:
Merge branch 'master' of https://github.com/microsoft/winget-pkgs into pr/felipecassiors/827

felipecassiors commented:
The current approach this script uses has been working for a while. I think it's ready to be merged, except that we have to change where the VCLibs `.AppX` packages are uploaded. We can use any reliable CDN, do you have any suggestion?

After merging, we can propose any further optimization, such as https://github.com/microsoft/winget-cli/issues/401.

denelon commented:
@felipecassiors we (@KevinLaMS and I) are also only holding on the VCLibs location. We're reaching out internally to see if we can find a good workable solution.

felipecassiors commented:
> @felipecassiors we (@KevinLaMS and I) are also only holding on the VCLibs location. We're reaching out internally to see if we can find a good workable solution. @zachcarp is pressing us to get this done.

Great to hear that. I think the best possible solution would be microsoft/winget-cli#401. I tried to take a look at the code to check how could I implement such this thing but did not understood, I'm not familiar with the Appx packaging and Visual Studio. But I'll give another try later.

felipecassiors commented:
I started some work at https://github.com/microsoft/winget-cli/issues/401 via https://github.com/microsoft/winget-cli/pull/456, but I'm not sure what I'm doing 😥.

Xeevis commented:
This script is great, thanks! Also when shortcut is added to SendTo menu, testing any manifest is just 2 clicks away.

**Path**: `shell:sendto`
**Target:** `pwsh "E:\GitHub\winget-pkgs\Tools\SandboxTest.ps1"`
**Icon:** `%windir%\system32\WindowsSandbox.exe`

![image](https://user-images.githubusercontent.com/5835044/86520885-ca0d9800-be49-11ea-9527-8bf3316b2c12.png)

felipecassiors commented:
> This script is great, thanks! Also when shortcut is added to SendTo menu, testing any manifest is just 2 clicks away.
> 
> **Path**: `shell:sendto`
> **Target:** `pwsh "E:\GitHub\winget-pkgs\Tools\SandboxTest.ps1"`
> **Icon:** `%windir%\system32\WindowsSandbox.exe`
> 
> ![image](https://user-images.githubusercontent.com/5835044/86520885-ca0d9800-be49-11ea-9527-8bf3316b2c12.png)

Wow, that's awesome! It becomes even handier.

zachcarp commented:
@felipecassiors - We have contacted the visual studio teams who maintains the VCLibs library in hopes of finding a long term solution for that package - but we don't have a confirmation on timeline yet. However, reading above you mentioned that you have this working without VCLibs? If thats the case, and you update the PR with it working without, we can get this merged now.

zachcarp commented:
@felipecassiors - Completely off topic, but could you share with us what tools your using for generating the gifs? They are quite helpful and I'm not familiar with how they are made.

felipecassiors commented:
> However, reading above you mentioned that you have this working without VCLibs? If thats the case, and you update the PR with it working without, we can get this merged now.

@zachcarp that's true, but it was with a manually built package via https://github.com/microsoft/winget-cli/pull/456. I did not use the official and released installer.

> @felipecassiors - Completely off topic, but could you share with us what tools your using for generating the gifs? They are quite helpful and I'm not familiar with how they are made.

For sure! With the awesome and [open source](https://github.com/ShareX/ShareX) [ShareX ](https://getsharex.com/) tool. It's available in the [Microsoft Store](https://www.microsoft.com/pt-br/p/sharex/9nblggh4z1sp). 

But of course, you can also:

```powershell
winget install sharex 
```

Then, for short, the default key shortcut to record a GIF is `Ctrl + Shift + Print Screen`.

felipecrs committed:
Merge branch 'master' of https://github.com/microsoft/winget-pkgs into sandbox-test

zachcarp commented:
@filipecassiors - Regarding this PR: I want to get this merged as it's a very useful tool. However, we are stuck until we can either get the script to work without the dependencies, or we can get them hosted at a more stable location (@denelon any input?). 

Thanks for the info regarding that tool, that could be very helpful for other issues. @denelon - Want to add a note about that tool in the templates for bugs/contributions/other?

felipecassiors commented:
For those who want to use this script before it gets merged, this is how I do:

```powershell
# Download the script
mkdir C:\SandboxTest\
Invoke-WebRequest -Uri 'https://raw.githubusercontent.com/felipecassiors/winget-pkgs/sandbox-test/Tools/SandboxTest.ps1' -OutFile C:\SandboxTest\SandboxTest.ps1
# Use it
C:\SandboxTest\SandboxTest.ps1 C:\Repos\winget-pkgs\manifests\GitHub\cli\0.10.1.yaml
```

felipecrs committed:
Fix wrong conflict resolution in README

denelon commented:
@zachcarp it looks like our best shot would be to host the dependencies in this repository until a better solution becomes available. I just pushed a new preview release of the client, and we will be flighting to insiders shortly. Let's look at this dependency issue next.

felipecrs committed:
Upgrade WinGet to v0.1.41821-preview

chausner commented:
> Good point. `winget validate` should indeed exit with a non-zero error code if validation fails (I created issue [microsoft/winget-cli#312](https://github.com/microsoft/winget-cli/issues/312)). For now, you could just parse the output and look for the string "Manifest validation succeeded.".

This is now fixed in master by https://github.com/microsoft/winget-cli/pull/535:
> Finally, the validate command now returns a non-zero value when there are issues. If there is a failure during validation, the value returned is 0x8A150029. If there are only warnings, the value returned is 0x8A150028

felipecassiors commented:
> > Finally, the validate command now returns a non-zero value when there are issues. If there is a failure during validation, the value returned is 0x8A150029. If there are only warnings, the value returned is 0x8A150028

Awesome, I'm going to update this PR as soon as the build gets released. I have some other new features to port back that I was working on at https://github.com/chocolatey-community/chocolatey-coreteampackages/pull/1512.

Trolldemorted commented:
Related question: How can we install winget inside the sandbox? Being able to use winget to provision sandboxes would be awesome.

felipecrs commented:
> Related question: How can we install winget inside the sandbox? Being able to use winget to provision sandboxes would be awesome.

This is what this script does

felipecrs committed:
Merge branch 'master' of https://github.com/microsoft/winget-pkgs into sandbox-test

Trolldemorted commented:
Thanks, now I get it! So
```
$desktopAppInstaller = @{
  fileName = 'Microsoft.DesktopAppInstaller_8wekyb3d8bbwe.appxbundle'
  url      = 'https://github.com/microsoft/winget-cli/releases/download/v0.1.41821-preview/Microsoft.DesktopAppInstaller_8wekyb3d8bbwe.appxbundle'
  hash     = '3fff593736c8272a640b5ba0e48343ccc0bbc1630b11abf057c22cedb6ac8ec1'
}
$vcLibs = @{
  fileName = 'Microsoft.VCLibs.140.00_14.0.27810.0_x64__8wekyb3d8bbwe.Appx'
  url      = 'https://github.com/felipecassiors/winget-pkgs/raw/da8548d90369eb8f69a4738dc1474caaffb58e12/Tools/SandboxTest_Temp/Microsoft.VCLibs.140.00_14.0.27810.0_x64__8wekyb3d8bbwe.Appx'
  hash     = 'fe660c46a3ff8462d9574902e735687e92eeb835f75ec462a41ef76b54ef13ed'
}
$vcLibsUwp = @{
  fileName = 'Microsoft.VCLibs.140.00.UWPDesktop_14.0.27810.0_x64__8wekyb3d8bbwe.Appx'
  url      = 'https://raw.githubusercontent.com/felipecassiors/winget-pkgs/da8548d90369eb8f69a4738dc1474caaffb58e12/Tools/SandboxTest_Temp/Microsoft.VCLibs.140.00.UWPDesktop_14.0.27810.0_x64__8wekyb3d8bbwe.Appx'
  hash     = '66de9fde9d2ebf18893a890987f35d2d145c18cc5ee0e8ecaa09477dcc13b16b'
}
$dependencies = @($desktopAppInstaller, $vcLibs, $vcLibsUwp)
foreach ($dependency in $dependencies) {

    # Only download if the file does not exist, or its hash does not match.
    if (-Not ((Test-Path -Path $dependency.fileName -PathType Leaf) -And $dependency.hash -eq $(get-filehash $dependency.fileName).Hash)) {
        # This downloads the file
        Write-Host "Downloading $($dependency.url) ..."
        try { 
            wget $dependency.url -o $dependency.fileName -UseBasicParsing
        } 
        catch {
            throw "Error downloading $($dependency.url) ."
        }
        if (-not ($dependency.hash -eq $(get-filehash $dependency.fileName).Hash)) {
            throw 'Hashes do not match, try gain.'
        }
    }
}

Add-AppxPackage -Path 'Microsoft.DesktopAppInstaller_8wekyb3d8bbwe.appxbundle' -DependencyPath 'Microsoft.VCLibs.140.00_14.0.27810.0_x64__8wekyb3d8bbwe.Appx','Microsoft.VCLibs.140.00.UWPDesktop_14.0.27810.0_x64__8wekyb3d8bbwe.Appx'
```
is currently the way to go to use winget inside the sandbox? I could not use DownloadFile inside the sandbox (no decent error message unfortunately), so I resorted to Invoke-WebRequest.

felipecrs committed:
Rewrite the whole script

Adding some cool features such as:
- Specify folder to mount (default to current)
- Make Manifest optional, so we can use the script to just start the
  Sandbox and install WinGet
- Add optional Script parameter, to specify which script block to run
  after all the steps inside the Sandbox
- Automatically refresh environment variables after installing the
  Manifest, and leave the function Update-Environment available for the
  users
- Use a temporary folder on system's temp directory, thus deleting the
  .gitignore
- Make Powershell open maximized and without splash messages
- Improve overall UX both in local machine and in the Sandbox
- Close Sandbox automatically if needed
- Upgrade WinGet to latest version

Most of the refactor came from my PR at
https://github.com/chocolatey-community/chocolatey-coreteampackages/pull/1512

felipecrs commented:
I rewrote the entire script and updated the PR description with a new GIF and some useful notes. Also, take a look at the [commit message](https://github.com/microsoft/winget-pkgs/pull/827/commits/e875fa93ac1706512d88d38c76b813ac1968da19) to get to know what was changed.

felipecrs commented:
@Trolldemorted you got it, but I wonder why don't you use this script instead. Take a look at the last changes, it should support most of the use cases now since it's now able to specify (or not) not only the Manifest to test but a custom script to run after and also the folder to mount in the Sandbox.

felipecrs committed:
Add tip for Update-Environment

msftrubengu commented on Tools/SandboxTest.ps1:
```diff
@@ -1,39 +1,66 @@
-  throw 'The Manifest file does not exist.'
+$ErrorActionPreference = "Stop"
+
+$mapFolder = [System.IO.Path]::GetFullPath($MapFolder)
```
I believe that GetFullPath will generate the path with the current directory where the PowerShell process started and will produce the wrong path if the user is running PowerShell, change directories and use the script.

I think it will be safer to use a combination of Get-Location + GetFullPath(String, String) with a prior call to IsPathFullyQualified to know if resolving the relative path is needed.

felipecrs replied:
It resolves the path in the situation you described. Since we're using `pwd`, it will always be right.

felipecrs replied:
The GIF in the PR description demonstrates that.

felipecrs replied:
NVM, it does not demonstrate that by this does:

![ZFpKotXBpO](https://user-images.githubusercontent.com/29582865/91506229-038ddd00-e8a8-11ea-979e-50e28532b3af.gif)

msftrubengu replied:
Yes, if `pwd` is used it works as expected. The problems comes when the user specifies MapFolder

![image](https://user-images.githubusercontent.com/36937303/91507041-961e8400-e888-11ea-8148-9e4166e78437.png)

felipecrs replied:
You are totally right. 

I think `Resolve-Path` should be enough, don't you think?

felipecrs replied:
Thanks for raising this, I just noticed that installing a manifest provided would not work if the path for it wasn't relative to the map folder. I fixed that by the way.

felipecrs committed:
Fix relative paths on map folder

felipecrs committed:
Copy the manifest instead of installing from the source

Trolldemorted commented:
@felipecrs I dislike having unneccessary "moving parts" (especially on my host) - my sandboxes need a powershell file which does the provisioning anyway, so having an additional one on the host would just split one problem into two.

felipecrs commented:
@Trolldemorted I see. It would be good to have such a script for general purposes published on Powershell Gallery.

Trolldemorted commented:
It would be good to have winget in the sandbox automatically without having to install it :)

felipecrs commented:
> It would be good to have winget in the sandbox automatically without having to install it :)

Winget was not even released and that's why we do not have it automatically on our computers. In my opinion there is no point of requesting such a thing before Winget gets installed by default on every Windows.

pbozzay commented:
> > It would be good to have winget in the sandbox automatically without having to install it :)
> 
> Winget was not released yet and that's why we do not have it automatically on our computers. In my opinion there is no point of requesting such a thing before Winget gets installed by default on every Windows.

Hey there, I'm a developer on Sandbox - this looks AWESOME. Sandbox doesn't have WinGet installed by default right now (as you know) so this workaround would be really useful for a lot of people. Any chance you'd be interested in creating a modified version of this to allow someone to start a sandbox with WinGet installed?

I would love to link to this (or a standalone fork of it) from https://github.com/microsoft/Windows-Sandbox-Utilities to share it with more users.

felipecrs committed:
Fix call without manifest param and enhance logs

felipecrs commented:
Hi @pbozzay, I'm glad to hear that you liked it.

> Any chance you'd be interested in creating a modified version of this to allow someone to start a sandbox with WinGet installed?

This is already possible (I just fixed a bug which was preventing it to work), give it a try with:
```sh-session
$ iwr -useb https://raw.githubusercontent.com/felipecassiors/winget-pkgs/sandbox-test/Tools/SandboxTest.ps1 | iex
```

Or download the script and run it normally, without any parameter.

Demo:

![AZqbNLAjue](https://user-images.githubusercontent.com/29582865/92961838-d3584980-f424-11ea-9a6a-d5331c911429.gif)

> I would love to link to this (or a standalone fork of it) from https://github.com/microsoft/Windows-Sandbox-Utilities to share it with more users.

Thanks for pointing this repository! I'll think about how could we link it to there.

felipecrs committed:
Merge branch 'master' of https://github.com/microsoft/winget-pkgs into sandbox-test

KevinLaMS reviewed (dismissed):
@felipecrs This is such an awesome script.  Thank you for authoring.
The Visual Studio team has published the APPX files for the runtime. 
https://docs.microsoft.com/en-us/troubleshoot/cpp/c-runtime-packages-desktop-bridge

The links to the files are as follows:
https://aka.ms/Microsoft.VCLibs.arm.14.00.Desktop.appx
https://aka.ms/Microsoft.VCLibs.arm64.14.00.Desktop.appx
https://aka.ms/Microsoft.VCLibs.x64.14.00.Desktop.appx
https://aka.ms/Microsoft.VCLibs.x86.14.00.Desktop.appx

If you update the script to use these, I am unblocked from approving!

Thanks
Kevin

felipecrs committed:
Use VCLibs UWPDesktop from official source and update winget

felipecrs committed:
Merge branch 'master' of https://github.com/microsoft/winget-pkgs into sandbox-test

felipecrs commented:
@KevinLaMS this is awesome!

I updated the PR, but sadly, they uploaded only one of the two dependencies that we need.

If we try to install `winget` only with the VCLibs that they provided (the UWPDesktop variant), the installation of winget fails with:

```log
Add-AppxPackage : Deployment failed with HRESULT: 0x80073CF3, Package failed updates, dependency or conflict validation.
Windows cannot install package Microsoft.DesktopAppInstaller_1.11.2521.0_x64__8wekyb3d8bbwe because this package depends on a framework that could
not be found. Provide the framework "Microsoft.VCLibs.140.00" published by "CN=Microsoft Corporation, O=Microsoft Corporation, L=Redmond,
S=Washington, C=US", with neutral or x64 processor architecture and minimum version 14.0.27810.0, along with this package to install. The frameworks
with name "Microsoft.VCLibs.140.00" currently installed are: {Microsoft.VCLibs
Windows cannot install package Microsoft.DesktopAppInstaller_1.11.2521.0_x64__8wekyb3d8bbwe because this package depends on a framework that could
not be found. Provide the framework "Microsoft.VCLibs.140.00" published by "CN=Microsoft Corporation, O=Microsoft Corporation, L=Redmond,
S=Washington, C=US", with neutral or x64 processor architecture and minimum version 14.0.27810.0, along with this package to install. The frameworks
with name "Microsoft.VCLibs.140.00" currently installed are: {Microsoft.VCLibs.140.00_14.0.27323.0_x64__8wekyb3d8bbwe}
NOTE: For additional information, look for [ActivityId] 5d7dc0bc-8886-0002-4b17-7e5d8688d601 in the Event Log or use the command line
Get-AppPackageLog -ActivityID 5d7dc0bc-8886-0002-4b17-7e5d8688d601
At C:\Users\WDAGUtilityAccount\Desktop\SandboxTest\Bootstrap.ps1:29 char:1
+ Add-AppxPackage -Path 'C:\Users\WDAGUtilityAccount\Desktop\SandboxTes ...
+ ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
    + CategoryInfo          : WriteError: (C:\Users\WDAGUt...bbwe.appxbundle:String) [Add-AppxPackage], IOException
    + FullyQualifiedErrorId : DeploymentError,Microsoft.Windows.Appx.PackageManager.Commands.AddAppxPackageCommand
```

However, I updated one of the dependencies to use the official source. Could you please ask the Visual Studio team to take a look at the other one?

KevinLaMS commented:
Sorry for the delay.  I was trying to repo, to menace Visual Studio.  But right now I am failing the script.  Any thoughts on what I am doing wrong?

![image](https://user-images.githubusercontent.com/2146880/95499010-83a06b80-0959-11eb-8eae-c1bc000495ec.png)

felipecrs committed:
Add some missing new lines

felipecrs committed:
Merge branch 'master' of https://github.com/microsoft/winget-pkgs into sandbox-test

felipecrs commented:
@KevinLaMS the issue should be fixed now. Can you please try again?

Xeevis commented:
@felipecrs Wouldn't it make sense for this awesome script to be put into it's own repository so it can be collaborated on and published as a public [PowerShell Gallery](https://www.powershellgallery.com/) module?

It would be much easier to consume, would cover wider use scenarios and can be updated with proper versioning. Usage could be something like:

```powershell
Install-Module -Name PSWingetSandbox -Scope CurrentUser # Install
Invoke-PSWingetSandbox -Manifest 'C:\winget-pkgs\manifests\Fork\Fork\1.56.1.yaml' # Install with manifest
Invoke-PSWingetSandbox -Name Microsoft.dotnet -Version 3.1.402 # Install older version with ID
```

I find myself often wanting to run/install some apps isolated so they don't pollute the OS and this would help greatly.

felipecrs committed:
Fetch latest WinGet release automatically

felipecrs committed:
Merge branch 'master' of https://github.com/microsoft/winget-pkgs into sandbox-test

felipecrs commented:
Since the WinGet CLI releases now have the SHA256 checksum available (ref https://github.com/microsoft/winget-cli/issues/568), now this script automatically fetches the latest (non-prerelease) version of WinGet automatically.

felipecrs commented:
> @felipecrs Wouldn't it make sense for this awesome script to be put into it's own repository so it can be collaborated on and published as a public [PowerShell Gallery](https://www.powershellgallery.com/) module?
> 
> It would be much easier to consume, would cover wider use scenarios and can be updated with proper versioning. Usage could be something like:
> 
> ```powershell
> Install-Module -Name PSWingetSandbox -Scope CurrentUser # Install
> Update-Module -Name PSWingetSandbox # Get latest version
> Invoke-PSWingetSandbox -Manifest 'C:\winget-pkgs\manifests\Fork\Fork\1.56.1.yaml' # Install with manifest
> Invoke-PSWingetSandbox -Name Microsoft.dotnet -Version 3.1.402 # Install older version with ID
> ```
> 
> I find myself often wanting to run/install some apps isolated so they don't pollute the OS and this would help greatly.

This is a great idea, and I plan to work on it.

felipecrs commented:
@KevinLaMS did you get any response from the VS team?

Xeevis commented:
ISSUE: Validation fails on warning

```powershell
PS> .\SandboxTest.ps1 E:\GitHub\winget-pkgs\manifests\TorProject\TorBrowser\10.0.9.yaml
--> Validating Manifest
Manifest validation succeeded with warnings.
Manifest Warning: Silent and SilentWithProgress switches are not specified for InstallerType exe. Please make sure the installer can run unattended.

Exception: E:\SandboxTest.ps1:31
Line |
  31 |      throw 'Manifest validation failed.'
     |      ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
     | Manifest validation failed.
```

felipecrs commented:
@Xeevis this is intended, we just rely on the status code returned by winget here. Providing a way to not fail on warnings would be a new CLI switch for winget itself, consider opening an Issue there if you think it's good.

felipecrs commented:
I'll think about adding a switch to skip the validation, though. For now, you can do something like:

```sh-session
.\SandboxTest.ps1 { winget install -m 10.0.9.yaml }
```

Or something like that, as long as 10.0.9 is in the current folder.

denelon commented:
@felipecrs have you been able to confirm if this works for the current version of the client with the new manifest schema? I see this PR has been open for a long time. It might be time to consider merging this one :)

felipecrs commented:
I'm going to try soon, @denelon.

felipecrs committed:
Merge branch 'master' of https://github.com/microsoft/winget-pkgs into sandbox-test

felipecrs committed:
Add -SkipManifestValidation switch and fix checksum

felipecrs committed:
Fix TLS issues, handle PowerShell 7, remove unneeded VCLibs dep and fix manifest as folder

felipecrs committed:
Restore file/folder check for manifest

felipecrs commented:
@denelon 

1. Added `-SkipManifestValidation` (https://github.com/microsoft/winget-pkgs/pull/827#issuecomment-774079583)
2. Fix TLS issues with `Invoke-WebRequest` and GitHub API.
3. Fix checksum file name convention in the winget-cli GitHub releases
4. Fix issues with PowerShell 7 in my tests
5. Remove the unneeded VCLibs dependency which was using my own URL (yay!)
6. Add compatibility with the new manifest schema (folders instead of files)

I think it's good to merge!

felipecrs committed:
Change checksum output filename

denelon commented:
@felipecrs how did you remove the dependency?

I'll ask one of the engineers to give this a review :)

felipecrs commented:
> how did you remove the dependency?

@denelon, well, it seems that the newest version of winget-cli does not depend on that specific `appx` anymore, I simply removed it, and it's working still.

palenshus commented on Tools/SandboxTest.ps1:
```diff
@@ -0,0 +1,298 @@
+
+Get-ChildItem $tempFolder -Recurse -Exclude $dependencies.fileName | Remove-Item -Force -Recurse
+
+if (-Not [String]::IsNullOrWhiteSpace($Manifest)) {
```
Aside from making this mandatory in the params, I'd check this once up top and fail out if it's whitespace, then you don't have to keep checking it

felipecrs replied:
Also depends on the other conversation, I guess.

felipecrs replied:
Also depends on the other conversation, I guess.

palenshus replied:
Leaving manifest optional is fine

palenshus commented on Tools/SandboxTest.ps1:
```diff
@@ -0,0 +1,298 @@
+
+$ProgressPreference = $oldProgressPreference
+
+$vcLibsUwp = @{
```
Maybe I misinterpreted [this comment](https://github.com/microsoft/winget-pkgs/pull/827#issuecomment-827807801), but I thought you don't need to install this anymore? (Forgive my ignorance, I don't actually work on winget-cli)

felipecrs replied:
In fact, this is what I don't need anymore:

https://github.com/microsoft/winget-pkgs/blob/cb06ad14ed3f82ca60a95f379398a901312c6f80/Tools/SandboxTest.ps1#L95-L99

felipecrs replied:
In fact, this is what I don't need anymore:

https://github.com/microsoft/winget-pkgs/blob/cb06ad14ed3f82ca60a95f379398a901312c6f80/Tools/SandboxTest.ps1#L95-L99

palenshus replied:
Got it thanks

palenshus commented on Tools/SandboxTest.ps1:
```diff
@@ -0,0 +1,298 @@
+
+# Create Bootstrap script
+
+# See: https://stackoverflow.com/a/14382047/12156188
```
I'd rename this for a little more clarity. Maybe Update-EnvironmentVariables?

palenshus replied:
Also, this implementation seems a little less fraught to me? https://stackoverflow.com/a/22670892/892770

felipecrs replied:
> Update-EnvironmentVariables

Well, it's really a long string, especially if you consider that it will be typed manually. But that's how PowerShell is, I agree with this change.

> Also, this implementation seems a little less fraught to me? https://stackoverflow.com/a/22670892/892770

Definitely. I don't know why I picked the first one, but I'll try and if it works I'll switch.

felipecrs replied:
> Update-EnvironmentVariables

Well, it's really a long string, especially if you consider that it will be typed manually. But that's how PowerShell is, I agree with this change.

> Also, this implementation seems a little less fraught to me? https://stackoverflow.com/a/22670892/892770

Definitely. I don't know why I picked the first one, but I'll try and if it works I'll switch.

palenshus commented on Tools/SandboxTest.ps1:
```diff
@@ -0,0 +1,298 @@
+
+
+
+
```
Why so much whitespace?

felipecrs replied:
Because of this (the progress bar stays on the screen even after the completion):

![U6ho8iV7OL](https://user-images.githubusercontent.com/29582865/90988322-56922800-e568-11ea-81f9-a32f191c35f9.gif)

But this is not happening anymore in my tests on 21H1. At the time the GIF were created I think I tested with 2004, so I don't know how it is for 20H2 (the current stable Windows version).

One option would be to hide the progress bar though, but I just wanted to make the script as transparent as possible. I even though about turning on the PowerShell debug feature to output each command being runned, but it turned out to be too noisy.

What do you think?

felipecrs replied:
Because of this (the progress bar stays on the screen even after the completion):

![U6ho8iV7OL](https://user-images.githubusercontent.com/29582865/90988322-56922800-e568-11ea-81f9-a32f191c35f9.gif)

But this is not happening anymore in my tests on 21H1. At the time the GIF were created I think I tested with 2004, so I don't know how it is for 20H2 (the current stable Windows version).

One option would be to hide the progress bar though, but I just wanted to make the script as transparent as possible. I even though about turning on the PowerShell debug feature to output each command being runned, but it turned out to be too noisy.

What do you think?

palenshus replied:
Gotcha, keep it then no prob

palenshus commented:
Very cool, this is super impressive! Just got a few comments here and there

felipecrs commented on README.md:
````diff
@@ -35,10 +35,20 @@ Although the Windows Package Manager YAML Generator can create YAML files with m
+
+Just provide the path to manifest as parameter:
+```powershell
+.\Tools\SandboxTest.ps1 <path-to-manifest>
````
@palenshus Well, my goal was to support the two kinds of manifests to keep backwards compatibility (it was files previously). Since you said you're not from the winget-cli team, it may be better to ask them about it. @denelon, any suggestion?

If yet we decide to keep accepting both files and folders, the help can be surely enhanced to clarify that.

felipecrs replied:
@palenshus Well, my goal was to support the two kinds of manifests to keep backwards compatibility (it was files previously). Since you said you're not from the winget-cli team, it may be better to ask them about it. @denelon, any suggestion?

If yet we decide to keep accepting both files and folders, the help can be surely enhanced to clarify that.

denelon replied:
If you have a singleton manifest you can point to the file, but the directory option works for both cases.

palenshus replied:
I see, I was only trying it on multi-file schemas. I'd just make it take a directory in both cases for consistency. Winget will validate the directory and ensure that if the manifestType is singleton, then there's only the one file in there

felipecrs commented on Tools/SandboxTest.ps1:
```diff
@@ -0,0 +1,298 @@
+
+Param(
+  [Parameter(Position = 0, HelpMessage = "The Manifest to install in the Sandbox.")]
+  [String] $Manifest,
```
The purpose of the script is to be flexible. Some times you just want to open sandbox, have winget installed, the current folder mounted and mess with it freely, this allows:

```sh-session
> .\Tools\SandboxTest.ps1 # works
> .\Tools\SandboxTest.ps1 -Script { winget install git; Update-Environment; git --version } # also works, no manifest
```

felipecrs replied:
The purpose of the script is to be flexible. Some times you just want to open sandbox, have winget installed, the current folder mounted and mess with it freely, this allows:

```sh-session
> .\Tools\SandboxTest.ps1 # works
> .\Tools\SandboxTest.ps1 -Script { winget install git; Update-Environment; git --version } # also works, no manifest
```

palenshus replied:
I see, I didn't realize that was an intended purpose, makes sense

felipecrs commented on Tools/SandboxTest.ps1:
```diff
@@ -0,0 +1,298 @@
+
+# Validate manifest file
+
+if (-Not $SkipManifestValidation -And -Not [String]::IsNullOrWhiteSpace($Manifest)) {
```
Thanks for the suggestion, but I kind of like the PowerShell verbosity. To me, it would not make sense use `!` instead of `-Not` while keeping `-And`.

felipecrs replied:
Thanks for the suggestion, but I kind of like the PowerShell verbosity. To me, it would not make sense use `!` instead of `-Not` while keeping `-And`.

felipecrs commented on Tools/SandboxTest.ps1:
```diff
@@ -0,0 +1,298 @@
+if (-Not $SkipManifestValidation -And -Not [String]::IsNullOrWhiteSpace($Manifest)) {
+  Write-Host '--> Validating Manifest'
+
+  if (-Not (Test-Path -Path $Manifest)) {
```
Depends on the other conversation also.

felipecrs replied:
Depends on the other conversation also.

felipecrs reviewed (commented):
Thank you so much for the comments, I'll work on them, but a few I still require some inputs.

felipecrs committed:
Update Update-EnvironmentVariables function

palenshus reviewed (commented):
Looks good to me
